docs: Add python API guide (#730)
nitrictech/docs: docs/guides/python/serverless-ai-api.mdx

Reviewer: This one looks like a good candidate for being featured. See https://github.com/nitrictech/docs/pull/664 on how to add it (image size etc).

- _Minimal configuration:_ The framework uses an “infrastructure from code” approach. We won't write separate infrastructure files; our Python code (and a tiny project config) is enough for Nitric to provision what's needed.

Now, let's get our tools ready and dive in.

## Prerequisites

Before we begin, make sure you have the following:

- **Python 3.8+** installed on your system (the latest 3.x is fine). Basic familiarity with Python is assumed.
- **Nitric CLI** installed - this is the command-line tool to create, run, and deploy Nitric projects. You can install it by following the [installation instructions](/get-started/installation).
- **uv** or **Pipenv** for managing Python dependencies. The Nitric Python templates use either uv or Pipenv.
- An **OpenAI API key** if you want to integrate with OpenAI's GPT for summarization.
- An **AWS account** (or Google Cloud or Azure account) with credentials configured on your machine, if you plan to deploy to the cloud. Nitric will use these credentials to provision resources. (For AWS, having the AWS CLI configured or environment vars set for access keys is enough.)  

>   _🔹 Note:_ You can deploy to any of the three major clouds - Nitric supports all. In this tutorial we'll mention AWS for concreteness, but feel free to use your preferred cloud.

Reviewer: This is broken, use the Note component

---
description: 'Build a Serverless AI-Powered Python API'
tags:
  - AI & Machine Learning
languages:
  - python
image: /docs/images/guides/serverless-ai-api/banner.png
image_alt: 'Serverless AI-Powered API Banner'
featured:
  image: /docs/images/guides/serverless-ai-api/banner.png
  image_alt: 'Serverless AI-Powered API Banner'

> published_at: 2025-04-03

Reviewer: ```suggestion
published_at: 2025-04-08
```

**Key Nitric benefits we'll see in action:**

- _Developer-friendly:_ Feels familiar if you've used web frameworks (you'll define routes and handlers in Python). No need to write Terraform or juggle AWS consoles - Nitric infers the necessary cloud resources from your code.
- _Accelerated development:_ Built-in local **development server** and dashboard for instant testing and debugging. You can run your serverless app locally with one command and even see a visual API explorer.
- _Portability:_ Nitric lets you **deploy anywhere** - AWS, GCP, Azure - by simply switching a setting, with all cloud-specific details handled for you. We'll deploy to AWS in this tutorial, but you could choose another cloud with minimal changes.
- _Minimal configuration:_ The framework uses an “infrastructure from code” approach. We won't write separate infrastructure files; our Python code (and a tiny project config) is enough for Nitric to provision what's needed.

Now, let's get our tools ready and dive in.

## Prerequisites

Before we begin, make sure you have the following:

> - **Python 3.8+** installed on your system (the latest 3.x is fine). Basic familiarity with Python is assumed.

Reviewer 2: ```suggestion
- **Python 3.11+** installed on your system (the latest 3.x is fine). Basic familiarity with Python is assumed.
```

- **Nitric CLI** installed - this is the command-line tool to create, run, and deploy Nitric projects. You can install it by following the [installation instructions](/get-started/installation).
- **uv** or **Pipenv** for managing Python dependencies. The Nitric Python templates use either uv or Pipenv.
- An **OpenAI API key** if you want to integrate with OpenAI's GPT for summarization.
- An **AWS account** (or Google Cloud or Azure account) with credentials configured on your machine, if you plan to deploy to the cloud. Nitric will use these credentials to provision resources. (For AWS, having the AWS CLI configured or environment vars set for access keys is enough.)

<Note>
  You can deploy to any of the three major clouds - Nitric supports all. In this
  tutorial we'll mention AWS for concreteness, but feel free to use your
  preferred cloud.
</Note>

> ## Project Setup - Your First Nitric App

Reviewer 2: Not sure if this tagline is required, isn't necessarily beginner-only content.

```suggestion
## Project Setup
```

Nitric projects have a simple structure: your code (services, functions, etc.) lives in a directory (by default called `services/`), and a `nitric.yaml` config file describes high-level settings (like which files are services and what runtime to use). We'll use the Nitric CLI to scaffold a new project:

1. **Create a new Nitric project:** Open a terminal and run:

```bash
nitric new my-summarizer py-starter
```

This will create a new folder `my-summarizer` with a Python starter template. (You can choose a different name for your project; `py-starter` selects the Python starter template.) If you omit the template, the CLI will interactively ask for a language/template. The CLI's `new` command guides you through project creation and offers ready-made templates.

If none of the templates suit your needs, you can always create your own.

2. **Install dependencies:** Navigate into the project and install the requirements. The template uses Pipenv, so you can do:

```bash
cd my-summarizer
uv sync
```

> This will install Nitric's Python SDK (and any other deps). If you used a different template, change this step to install dependencies accordingly.

Reviewer 2: ```suggestion
This will install Nitric's Python SDK (and any other dependencies). If you used a different template, change this step to install dependencies accordingly.
```

3. **Explore the scaffold:** The generated project should have a structure like:

```
my-summarizer/
├── services/
│   └── api.py
├── nitric.yaml
├── pyproject.toml
├── python.dockerfile
├── python.dockerfile.dockerignore
└── README.md
```

The `services/api.py` is a sample service that Nitric created (a simple “Hello, World” HTTP endpoint). We'll replace this with our own code. The `nitric.yaml` config by default tells Nitric to treat any `*.py` in `services/` as a service and what runtime to use. We'll keep the config mostly default.

> 4. **(Optional) Verify the template works:** You can test the starter project right away by running `nitric start` inside the project. This spins up Nitric's local development server. Check the terminal output - you should see that an API endpoint (perhaps `/hello/:name`) is available locally at some port (like 4001). For now, stop the server (Ctrl+C) - we're going to write our own service next.

Reviewer 2: ```suggestion
4. **(Optional) Verify the template works:** You can test the starter project right away by running `nitric start` inside the project. This spins up Nitric's local development server. Check the terminal output - you should see that an API endpoint is available locally at some port (like 4001). For now, stop the server (Ctrl+C) - we're going to write our own service next.
```

With the project set up, let's start coding our API.

## Writing the Python API Service (with AI Integration)

Our goal is to create an HTTP endpoint `/summarize` (for example) that accepts some text and returns a summary of that text. We'll implement this as an **API service** in Nitric. In Nitric's model, an **API** can have multiple RESTful routes, served by one or more services. We define an API and then use decorators to attach function handlers to specific HTTP methods and paths - very similar to frameworks like Flask or FastAPI, but with Nitric's `@api` decorators.

**Steps:**

1. **Create a new service file:** In the `services/` directory, create a file `summarize.py` (you can name it differently, but let's go with this for clarity).

2. **Import Nitric and dependencies:** We'll need Nitric's API utilities and the OpenAI library (for AI). Make sure you add `openai` to your dependencies (`uv add openai`). Then, in `summarize.py`:

> ```python

Reviewer 2: ````suggestion
```python title:services/summarize.py
````

import os
from openai import OpenAI
from nitric.resources import api
from nitric.application import Nitric
from nitric.context import HttpContext
```

Here we import `api` to create an API resource, `Nitric` to run the app, and `HttpContext` for typing (the context object passed into handlers). We'll use `os` to read the OpenAI API key from an environment variable, and `openai` to call the OpenAI service.

3. **Configure OpenAI API key:** It's best not to hardcode secrets. If you have your OpenAI API key, set it as an environment variable (e.g., create a .env file for the project containing `OPENAI_API_KEY=sk-...`). We can then configure the OpenAI client in code:

> ```python

Reviewer 2: ````suggestion
```python title:services/summarize.py
````

# Configure OpenAI API key from env
api_key = os.environ.get("OPENAI_API_KEY")
if not api_key:
  raise Exception("Please set the OPENAI_API_KEY environment variable")
```

This will fetch the API key from your environment. (If you prefer, Nitric also has a built-in **Secrets** feature to securely manage secrets, but for simplicity we'll stick to an env var here.)

4. **Create the OpenAI client:** Now, use Nitric to create a POST route for summarization:

> ```python

Reviewer 2: ````suggestion
```python title:services/summarize.py
````

# Create OpenAI client
client = OpenAI(api_key=api_key)

> ```

Reviewer 3: Could also provide the user the option of testing locally using ollama if they have it installed

5. **Define the API and route handler:** Now, use Nitric to create a POST route for summarization:

> ```python

Reviewer 2: ````suggestion
```python title:services/summarize.py
````

# Create a Nitric API named 'main'
summarize_api = api("main")

@summarize_api.post("/summarize")
async def summarize_text(ctx: HttpContext) -> None:
  # Extract input text from request (assuming JSON body with a 'text' field)
  req_data = ctx.req.json
  if not req_data or 'text' not in req_data:
    ctx.res.status = 400
    ctx.res.body = {"error": "No text provided for summarization."}
    return

  text_to_summarize = req_data['text']
  # Call OpenAI API to get summary (this is a blocking call, for demo simplicity)
  try:
    response = client.chat.completions.create(
      model="gpt-3.5-turbo",
      messages=[{"role": "user", "content": f"Summarize the following text:\n\n{text_to_summarize}"}],
      max_tokens=50
    )
    # Extract the assistant's reply (summary text)
    summary = response.choices[0].message.content
    # Set response
    ctx.res.body = {"summary": summary.strip()}
  except Exception as e:
    ctx.res.status = 500
    ctx.res.body = {"error": str(e)}

> Nitric.run()

Reviewer 2: Theres a section that adds this below
```suggestion
```

- The OpenAI response contains a list of choices; we extract the content of the first choice (which is the model's summary text).
- We set `ctx.res.body` to a JSON object containing the summary. Nitric will serialize this to JSON for the HTTP response. By default, returning a Python dict as `res.body` results in a JSON response.
- If the OpenAI API call fails (e.g., network issue or invalid key), we catch the exception and return a 500 error with the message.

_A note on async:_ We defined the handler as `async def`. The Nitric framework supports async handlers, which is great for I/O operations. The OpenAI Python SDK call we used (`ChatCompletion.create`) is a blocking call. In a real app, you might integrate an async HTTP client or use OpenAI's async methods (if available) to avoid blocking the event loop. However, for simplicity in this tutorial, this is acceptable - our focus is on the integration, not optimizing async usage.

5. **Finalize the Nitric app run:** After defining all your routes and handlers in the file, make sure to add:

> ```python

Reviewer 2: ````suggestion
```python title:services/summarize.py
````

Nitric.run()
```

This call tells Nitric to start the application (it will scan for all defined services/routes and run the local server when invoked via `nitric start`). In our case, it picks up `summarize_api` and the attached route.

That's it for our service code! We have a complete Python backend that will accept text and return an AI-generated summary. The entire `summarize.py` file should look like this (for reference):

> ```python {{ title: services/summarize.py }}

Reviewer 2: ````suggestion
```python title:services/summarize.py
````

import os
from openai import OpenAI
from nitric.resources import api
from nitric.application import Nitric
from nitric.context import HttpContext

# Configure OpenAI API key from env
api_key = os.environ.get("OPENAI_API_KEY")
if not api_key:
  raise Exception("Please set the OPENAI_API_KEY environment variable")

# Create OpenAI client
client = OpenAI(api_key=api_key)

# Create a Nitric API named 'main'
summarize_api = api("main")

@summarize_api.post("/summarize")
async def summarize_text(ctx: HttpContext) -> None:
  # Extract input text from request (assuming JSON body with a 'text' field)
  req_data = ctx.req.json
  if not req_data or 'text' not in req_data:
    ctx.res.status = 400
    ctx.res.body = {"error": "No text provided for summarization."}
    return

  text_to_summarize = req_data['text']
  # Call OpenAI API to get summary (this is a blocking call, for demo simplicity)
  try:
    response = client.chat.completions.create(

>       model="gpt-3.5-turbo",

Reviewer 3: should we allow this be to be configurable?

Reviewer 4: suggest we either add a note or extend this as a future improvement.

      messages=[{"role": "user", "content": f"Summarize the following text:\n\n{text_to_summarize}"}],

>       max_tokens=50

Reviewer 3: This seems very low, has this been tested?

Reviewer 4: Yup, tested and works.

Local testing is super fast - you can edit your Python code, save, and Nitric auto-reloads the changes (the dashboard and local server update on the fly). This tight feedback loop means you can iterate on your function logic quickly. Once you're satisfied with local results, it's time to deploy our app to the cloud 🚀.

## Deploying to the Cloud

Deploying a serverless app usually involves setting up cloud services (like an API gateway, a function, IAM permissions, etc.) which can be a lot of work manually. Nitric automates **all of this** for us. With a couple of CLI commands, Nitric will provision the necessary resources on your chosen cloud provider and get our API running live.

**1. Create a deployable stack:** In Nitric, a _stack_ is basically a deployment environment (combining your app code with cloud config like region and provider). We need to create one before our first deploy. Run:

```bash
nitric stack new prod-stack
```

_(You can name the stack as you like, e.g., “dev-stack” or “prod-stack”.)_

The CLI will prompt you a cloud provider (choose “AWS” for this tutorial, or GCP/Azure if you prefer),

> This command creates a stack configuration file (for example, `stacks/prod-stack.yaml`) in your project. Open this file - you'll see it lists the chosen provider and region, and some default config for how to deploy your services. Nitric has already made best-practice choices (like naming conventions, runtime settings, etc.) which you can tweak if needed. Ensure the region is set correctly for your account.

Reviewer 2: ```suggestion
This command creates a stack configuration file (for example, `prod-stack.yaml`) in your project. Open this file - you'll see it lists the chosen provider, an empty region key, and some default config for how to deploy your services. Nitric has already made best-practice choices (like naming conventions, runtime settings, etc.) which you can tweak if needed. Ensure the region is set correctly for your account.
```

**2. Deploy the stack:** Now the magic moment. Run:

```bash
nitric up
```

> This tells Nitric to deploy (“up”) the stack we just created. The CLI will now translate your application into cloud resources and deploy them. Under the hood, Nitric uses cloud IaC providers (like Pulumi or Terraform) to create the infra from your code definitions, but you don't have to write any of that yourself. It will automatically set up everything: an HTTP endpoint (API Gateway or equivalent), a cloud function to run your `summarize_text` code, and any needed permissions (IAM roles, etc.).

Reviewer 2: ```suggestion
This tells Nitric to deploy (“up”) the stack we just created. The CLI will now translate your application into cloud resources and deploy them. Under the hood, Nitric uses cloud IaC providers (like Pulumi or Terraform) to create the infra from your code definitions, but you don't have to write any of that yourself. It will automatically set up everything: an HTTP endpoint (API Gateway or equivalent), a cloud function to run your `summarize_text` code, and any required permissions (IAM roles, etc.).
```

You'll see a progress log in the console. The first deployment may take a few minutes as it's creating cloud resources. Once complete, Nitric will output the **endpoint URL** for your API. It might look something like:

```
https://xxxxxx.execute-api.us-east-1.amazonaws.com
```

(The exact format depends on the cloud - for AWS, it uses API Gateway; for others, a similar endpoint will be provided.

> **3. Test the live endpoint:** Time to test our deployed function on the cloud! You can use curl again, but now replace the localhost URL with the one given by Nitric. For example:

Reviewer 2: ```suggestion
**3. Test the live endpoint:** Time to test our deployed function on the cloud! You can use `curl` again, but now replace the localhost URL with the one given by Nitric. For example:
```

```bash
curl -X POST https://<<your-api-url>>/summarize \
  -H "Content-Type: application/json" \
  -d '{"text": "OpenAI API integration with serverless functions is very powerful, allowing applications to leverage AI without managing infrastructure."}'
```

> _(Make sure to put your actual endpoint URL in place of `<<your-api-url>>`._

Reviewer 2: ```suggestion
_Make sure to put your actual endpoint URL in place of `<<your-api-url>>`._
```

> You should get a JSON response with a summary, similar to what you saw locally. Congratulations - your AI-powered summarization service is now live on the internet, running serverlessly! 🎉

Reviewer 2: Just a suggestion

```suggestion
You should get a JSON response with a summary, similar to what you saw locally. Congratulations - your AI-powered summarization service is now live on the internet, running completely on serverless infrastructure! 🎉
```

- Add other Nitric resources: need a [database](/sql) or [key-value store](/keyvalue) to cache results? Nitric offers `sql()` and `kv()` resources that can be used similarly to how we defined the API. Want to trigger actions on a schedule (cron jobs) or handle file storage in the cloud? Nitric has primitives for [schedules](/schedules), [buckets](/storage) (object storage), [queues](/messaging#queues), pub-sub [topics](/messaging#topics), and more - all accessible through simple code constructs, and all cloud-agnostic. (For example, you could schedule a daily job to fetch some data and summarize it, using Nitric's `schedule` resource.)
- Deploy to another cloud: Try deploying the same app to Azure or GCP. It should be as easy as creating a new stack for that provider. This can be a great way to learn differences in cloud offerings without changing your code - Nitric will map your API to the equivalent service (e.g., Azure Container Apps + API Management, or GCP Cloud Run + API Gateway).

> As you continue exploring, check out the official Nitric documentation and examples for deeper dives. The [Nitric docs site](https://nitric.io/docs) has guides on all resource types and features (for instance, how to use **Queues/Topics for event-driven patterns**, or how to secure your API with auth). There's also a community [Discord](https://nitric.io/chat) and more examples on GitHub if you run into questions.

Reviewer 2: ```suggestion
As you continue exploring, check out the official Nitric documentation and examples for deeper dives. The [Nitric docs site](https://nitric.io/docs) has guides on all resource types and features (for instance, how to use **Queues/Topics for event-driven patterns**, or how to secure your API with auth). There's also a community [Discord](https://nitric.io/chat) and more [examples](https://github.com/nitrictech/examples) on GitHub if you want inspiration.
```

---
description: 'Build a Serverless AI-Powered Python API'
tags:
  - AI & Machine Learning
languages:
  - python
image: /docs/images/guides/serverless-ai-api/banner.png
image_alt: 'Serverless AI-Powered API Banner'
featured:
  image: /docs/images/guides/serverless-ai-api/banner.png
  image_alt: 'Serverless AI-Powered API Banner'

> published_at: 2025-04-08

Reviewer: ```suggestion
published_at: 2025-05-08
```

> published_at: 2025-05-08

Reviewer: ```suggestion
published_at: 2025-05-09
```
